shell/packages/sandstorm-ui-powerbox/powerbox-server.js

```diff
@@ -31,8 +68,240 @@ Meteor.methods({
       { userIsAdmin: Meteor.userId() },
     ];
 
-    // TODO: refactor: reaches out into core.js
-    const sturdyRef = waitPromise(saveFrontendRef(frontendRefVariety, apiTokenOwner, requirements)).sturdyRef;
-    return sturdyRef.toString();
+    // TODO(cleanup): refactor: reaches out into core.js
+    const sturdyRef = waitPromise(saveFrontendRef(frontendRefVariety, apiTokenOwner, requirements)).sturdyRef.toString();
+    return { sturdyRef, descriptor };
+  },
+});
+
+class PowerboxOption {
+  constructor(fields) {
+    _.extend(this, fields);
+  }
+
+  intersect(other) {
+    // Intersect two options with the same ID. Used when combining matches from multiple tags
+    // in the same descriptor. The tags are a conjunction.
+    //
+    // Returns true if there was any overlap, or false if there was no overlap and therefore the
+    // option should be dropped.
+
+    if (other._id != this._id) {
+      throw new Error("can only merge options with the same ID");
+    }
+
+    if (other.grainId) {
+      if (this.uiView && !other.uiView) delete this.uiView;
+      if (this.hostedObject && !other.hostedObject) delete this.hostedObject;
+      return !!(this.uiView || this.hostedObject);
+    } else {
+      // No intersection logic needed for other types.
+      return true;
+    }
+  }
+
+  union(other) {
+    // Union two options with the same ID. Used when combining matches from multiple descriptors.
+    // The descriptors are a disjunction.
+
+    if (other._id != this._id) {
+      throw new Error("can only merge options with the same ID");
+    }
+
+    if (other.grainId) {
+      if (!this.uiView && other.uiView) this.uiView = other.uiView;
+      if (!this.hostedObject && other.hostedObject) this.hostedObject = other.hostedObject;
+    }
+  }
+
+  subtract(other) {
+    // Remove `other` from this. Used when `other` is a match deemed unacceptable. Returns true
+    // if there's anything left, false otherwise.
+
+    if (other._id != this._id) {
+      throw new Error("can only merge options with the same ID");
+    }
+
+    if (other.grainId) {
+      if (this.uiView && other.uiView) delete this.uiView;
+      if (this.hostedObject && other.hostedObject) delete this.hostedObject;
+      return !!(this.uiView || this.hostedObject);
+    } else {
+      return false;
+    }
+  }
+}
+
+const specialCaseTypes = {
+  // This object maps tag IDs to functions which return lists of matches for that tag.
+
+  15831515641881813735(db, userId, value) {  // UiView
```

Comment:
Oh wow, because if that value is treated as a number, we lose precision because we only have 53-bit numerics.  Amazing.

That might be worth filing a bug against JSCS.